functional_programming_tests/test/without_functional_programming.cpp

```diff
@@ -0,0 +1,205 @@
+#include <gtest/gtest.h>
+
+#include <cmath>
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <vector>
+
+#include <rclcpp/rclcpp.hpp>
+#include <std_msgs/msg/u_int8_multi_array.hpp>
+
+using namespace std::chrono_literals;
+
+// production_code.h/cc
+template<typename T>
+using Map = std::vector<std::vector<T>>;
+
+// X, Y position
+struct Pos {
+	size_t x;
+	size_t y;
+};
+
+using Path = std::vector<Pos>;
+
+// Operator overload to print path
+std::ostream& operator<<(std::ostream& os, std::vector<Pos> const& path) {
+	for (auto const& pose:path) {
+		os << "(" << pose.x << ", " << pose.y << ")\n";
+	}
+	return os;
+}
+
+// Operator overload for position comparison
+bool operator==(Pos const& lhs, Pos const& rhs) {
+	return lhs.x == rhs.x && lhs.y == rhs.y;
+}
+
+class PathGenerator {
+public:
+	PathGenerator() : node_(std::make_shared<rclcpp::Node>("bad_test_subscriber")), executor_(std::make_shared<rclcpp::executors::SingleThreadedExecutor>()) {
+        // Create ROS2 subscriber for the costmap
+		costmap_subscriber_ = node_->create_subscription<std_msgs::msg::UInt8MultiArray>("/test/costmap", rclcpp::SystemDefaultsQoS(), [this](const std_msgs::msg::UInt8MultiArray::SharedPtr msg){costmap_callback(msg);});
+
+		// Let the node spin to execute get_costmap callbacks
+		executor_->add_node(node_);
+		executor_thread_ = std::thread([this]() { executor_->spin(); });    		
+	}
+	~PathGenerator() {
+		executor_->cancel();
+		executor_thread_.join();		
+	}
+    std::optional<Path> generate_global_path(Pos const& start, Pos const& goal);
+    void costmap_callback(const std_msgs::msg::UInt8MultiArray::SharedPtr msg);
+
+    Map<unsigned char> get_costmap() const { return map_;}
+private:
+	rclcpp::Node::SharedPtr node_;
```

Comment:
The tutorial would have you inherit from `Node` instead of compose with it. I obviously prefer composition but it does defocus the point of your example a little.
I think I would prefer this to be in a different example of composition versus inheritance

Reply:
Done